### PR TITLE
Correct the type of DispatchOptions["headers"]

### LINKED ISSUE #1892: The type of set-cookies obtained by heaers is inconsistent with the actual type?
## Bug Description

`headers["set-cookie"]`returned is a string, type declaration is a `string [] | undefined`?

```js
 const { headers } = await request(index_url);
 console.log(headers);
/* {
  date: 'Sat, 28 Jan 2023 09:43:48 GMT',
  'content-type': 'text/html',
  'content-length': '2920',
  connection: 'keep-alive',
  vary: 'Accept-Encoding',
  'last-modified': 'Sat, 17 Sep 2022 08:52:52 GMT',
  'accept-ranges': 'bytes',
  etag: '"1d8ca72dec58168"',
  'set-cookie': 'SERVERID=f7a91655448e700d4049e6f0ca2b6b52|1674899028|1674899028;Path=/'
} */
console.log(headers["set-cookie"]);
/*
SERVERID=f7a91655448e700d4049e6f0ca2b6b52|1674899028|1674899028;Path=/
*/
```

### PR DESCRIPTION
**Pull request by pan93412**
## This relates to...

Fixed #1892 
The simpler solution of #1893

## Rationale

According to core/util@parseHeaders, the current behavior is:

- By default, the entry type is a `string`.
- The entry type turns to a `string[]` when there are duplicated entries.

This behavior is not as same as what `node:http` does currently; therefore, `IncomingHttpHeaders` can't reflect our parsed data.

## Changes

image

This commit attempts to reflect this parsing behavior by redefining it to `Record<string, string | string[]>`.

### Features

N/A

### Bug Fixes

See Changes.

### Breaking Changes and Deprecations

- The old workaround to this wrong declaration should be removed now.

## Status

KEY: S = Skipped, x = complete

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] ~~Benchmarked (**optional**)~~
- [x] Documented
- [x] Review ready
- [ ] In review
- [ ] Merge ready

Note that I can't pass `npm run test`. Seems not related:

image

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by pan93412** [comment]
request for review: @mcollina @ronag

**Review by mcollina — APPROVED** [review]
lgtm

File: types/header.d.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/**
+ * The header type declaration of `undici`.
+ */
+export type IncomingHttpHeaders = Record<string, string | string[]>;
```
Comment (SimenB) [inline types/header.d.ts:4]: any particular reason why `undefined` is not allowed? Now it's a type error to pass `headers` from an incoming http request into `undici`

Comment (pan93412) [reply types/header.d.ts:4]: I think it is just a mistake – no particular reason there. Maybe we can open another PR to fix it?

Comment (SimenB) [reply types/header.d.ts:4]: cool 👍 I personally can't think of a reason to disallow `undefined` at least

**Pull request merged**